File: src/logical/write.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES. All rights
 * reserved. SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gqe/logical/write.hpp>

#include <gqe/logical/utility.hpp>
#include <gqe/utility/helpers.hpp>

#include <sstream>

namespace gqe {

namespace logical {

write_relation::write_relation(std::shared_ptr<relation> input_relation,
                               std::vector<std::string> column_names,
                               std::vector<cudf::data_type> column_types,
                               std::string table_name)
  : relation({std::move(input_relation)}, {}),
    _column_names(std::move(column_names)),
    _column_types(std::move(column_types)),
    _table_name(std::move(table_name))
{
}

relation::relation_type write_relation::type() const noexcept { return relation_type::write; }

std::vector<cudf::data_type> write_relation::data_types() const
{
  auto const child_relation = children_unsafe();
  assert(child_relation.size() == 1);
  auto const child_types = child_relation[0]->data_types();

  assert(_column_types == child_types);

  return _column_types;
}

std::string write_relation::to_string() const
{
  std::ostringstream ss;
  ss << "{\"Write\" : {\n"
     << "\t\"children\" : " << utility::list_to_string(children_unsafe()) << "\n"
     << "}}";

  return ss.str();
}

std::string write_relation::table_name() const { return _table_name; }

std::vector<std::string> write_relation::column_names() const { return _column_names; }

bool write_relation::operator==(const relation& other) const
{
  auto this_type = this->type();
  if (this_type != other.type()) {
    utility::log_relation_comparison_message(
      this_type,
      "operator==() relation type mismatch with " + utility::relation_type_str(other.type()));
    return false;
  }
  auto other_write_relation = dynamic_cast<const write_relation*>(&other);
  // Compare attributes
  if (this->column_names() != other_write_relation->column_names()) {
    utility::log_relation_comparison_message(this_type, "operator==(): column names mismatch");
    return false;
  }
  if (this->table_name() != other_write_relation->table_name()) {
    utility::log_relation_comparison_message(this_type, "operator==(): table names mismatch");
    return false;
  }
  if (this->data_types() != other_write_relation->data_types()) {
    utility::log_relation_comparison_message(this_type, "operator==(): data types mismatch");
    return false;
  }
  // Compare members defined in base class
  return relation::compare_relation_members(other);
}

}  // namespace logical
}  // namespace gqe

File: rust/gqe-cli/src/substrait_ext/write.rs
// Copyright 2026 NVIDIA Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Substrait encoding and decoding for Write relations (COPY FROM).

use arrow::datatypes::Schema;
use prost::Message;
use substrait::proto::read_rel::local_files::FileOrFiles;
use substrait::proto::read_rel::local_files::file_or_files::{
    FileFormat, ParquetReadOptions, PathType,
};
use substrait::proto::read_rel::{self, LocalFiles};
use substrait::proto::write_rel::{OutputMode, WriteOp, WriteType};
use substrait::proto::{
    self, NamedObjectWrite, Plan, PlanRel, ReadRel, Rel, WriteRel, plan_rel, rel,
};

use super::types::schema_to_named_struct;

/// Encode a COPY FROM statement as a serialized Substrait `Plan`.
pub fn encode_copy_from(table_name: &str, file_path: &str, schema: &Schema) -> Vec<u8> {
    let named_struct = schema_to_named_struct(schema);

    let read = ReadRel {
        common: None,
        base_schema: Some(named_struct.clone()),
        filter: None,
        best_effort_filter: None,
        projection: None,
        advanced_extension: None,
        read_type: Some(read_rel::ReadType::LocalFiles(LocalFiles {
            items: vec![FileOrFiles {
                partition_index: 0,
                start: 0,
                length: 0,
                path_type: Some(PathType::UriPath(file_path.to_string())),
                file_format: Some(FileFormat::Parquet(ParquetReadOptions {})),
            }],
            advanced_extension: None,
        })),
    };

    let write = WriteRel {
        write_type: Some(WriteType::NamedTable(NamedObjectWrite {
            names: vec![table_name.to_string()],
            advanced_extension: None,
        })),
        table_schema: Some(named_struct),
        op: WriteOp::Insert as i32,
        input: Some(Box::new(Rel {
            rel_type: Some(rel::RelType::Read(Box::new(read))),
        })),
        output: OutputMode::NoOutput as i32,
        common: None,
        advanced_extension: None,
        create_mode: 0,
    };

    let plan = Plan {
        relations: vec![PlanRel {
            rel_type: Some(plan_rel::RelType::Rel(Rel {
                rel_type: Some(rel::RelType::Write(Box::new(write))),
            })),
        }],
        ..Default::default()
    };

    let mut buf = Vec::new();
    plan.encode(&mut buf)
        .expect("Substrait plan encoding failed");
    buf
}

/// Decoded contents of a Substrait `WriteRel` (COPY FROM).
pub struct DecodedCopyFrom {
    pub table_name: String,
    pub file_path: String,
}

/// Decode a Substrait `WriteRel` into its components.
///
/// Currently only supports INSERT with a `ReadRel(LocalFiles)` input
/// (i.e. COPY FROM a local parquet file/directory).
pub fn decode_copy_from(write: &proto::WriteRel) -> Result<DecodedCopyFrom, String> {
    let op = WriteOp::try_from(write.op).map_err(|_| format!("Unknown write op: {}", write.op))?;
    if op != WriteOp::Insert {
        return Err(format!("Write operation not supported: {op:?}"));
    }

    let table_name = match &write.write_type {
        Some(WriteType::NamedTable(named)) => named
            .names
            .last()
            .ok_or("WriteRel NamedTable has no names")?
            .clone(),
        _ => return Err("WriteRel has no named table write type".into()),
    };

    let input = write.input.as_ref().ok_or("WriteRel has no input")?;

    let file_path = match &input.rel_type {
        Some(rel::RelType::Read(read_rel)) => match &read_rel.read_type {
            Some(read_rel::ReadType::LocalFiles(local_files)) => {
                let item = local_files.items.first().ok_or("LocalFiles has no items")?;
                match &item.path_type {
                    Some(PathType::UriPath(p) | PathType::UriFile(p) | PathType::UriFolder(p)) => {
                        p.clone()
                    }
                    other => return Err(format!("Unsupported path type: {other:?}")),
                }
            }
            other => return Err(format!("Expected LocalFiles read type, got: {other:?}")),
        },
        other => {
            return Err(format!(
                "Expected ReadRel input for WriteRel, got: {other:?}"
            ));
        }
    };

    Ok(DecodedCopyFrom {
        table_name,
        file_path,
    })
}
